File: services/nodejs-microservices/order-service/src/domain/models/Order.ts
export interface Order {
  id: string;
  userId: string;
  status: OrderStatus;
  totalAmount: number;
  shippingAddressId?: string;
  paymentId?: string;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

export enum OrderStatus {
  PENDING = 'PENDING',
  INVENTORY_RESERVED = 'INVENTORY_RESERVED',
  PAYMENT_PROCESSING = 'PAYMENT_PROCESSING',
  CONFIRMED = 'CONFIRMED',
  CANCELLED = 'CANCELLED',
  REFUNDED = 'REFUNDED',
  FAILED = 'FAILED',
}

export interface OrderItem {
  id: string;
  orderId: string;
  productId: string;
  productName?: string;
  quantity: number;
  unitPrice: number;
  subtotal: number;
}

/**
 * Saga State tracks the progress of a distributed transaction.
 * If any step fails, we use compensation data to rollback previous steps.
 */
export interface SagaState {
  id: string;
  orderId: string;
  currentStep: SagaStep;
  status: SagaStatus;
  compensationData?: any;
  createdAt: Date;
  updatedAt: Date;
}

export enum SagaStep {
  VALIDATE_CART = 'VALIDATE_CART',
  RESERVE_INVENTORY = 'RESERVE_INVENTORY',
  PROCESS_PAYMENT = 'PROCESS_PAYMENT',
  CONFIRM_ORDER = 'CONFIRM_ORDER',
  COMPLETED = 'COMPLETED',
}

export enum SagaStatus {
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  COMPENSATING = 'COMPENSATING',
  FAILED = 'FAILED',
}

File: services/dotnet-microservices/OrderService/Domain/Models/Order.cs
namespace OrderService.Domain.Models;

public record Order(
    Guid Id,
    Guid UserId,
    string Status,
    decimal TotalAmount,
    string? ShippingAddress,
    string SagaState,
    List<OrderItem> Items,
    DateTime CreatedAt,
    DateTime UpdatedAt
);

public record OrderItem(
    Guid Id,
    Guid OrderId,
    Guid ProductId,
    string ProductName,
    decimal UnitPrice,
    int Quantity
);

public record CreateOrderRequest(
    Guid UserId,
    string? ShippingAddress,
    List<OrderItemRequest> Items
);

public record OrderItemRequest(
    Guid ProductId,
    string ProductName,
    decimal UnitPrice,
    int Quantity
);

// Saga states
public static class SagaStates
{
    public const string Initiated = "initiated";
    public const string InventoryReserved = "inventory_reserved";
    public const string PaymentProcessed = "payment_processed";
    public const string Confirmed = "confirmed";
    public const string Failed = "failed";
    public const string Compensating = "compensating";
    public const string Cancelled = "cancelled";
}

public static class OrderStatuses
{
    public const string Pending = "pending";
    public const string Confirmed = "confirmed";
    public const string Shipped = "shipped";
    public const string Delivered = "delivered";
    public const string Cancelled = "cancelled";
}
